compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -533,33 +535,49 @@ abstract class PrepJSInterop extends plugins.PluginComponent
         } else if (enclosingOwner is OwnerKind.JSMod) {
           reporter.error(implDef.pos, "Scala.js-defined JS objects " +
               "may not have inner native JS classes or objects")
-        } else if (!sym.isTrait && (enclosingOwner is OwnerKind.JSNativeMod)) {
-          /* Store the fully qualified JS name in an explicit @JSFullName
-           * annotation, before `flatten` destroys the name and (in 2.10) the
-           * original owner chain.
+        } else if (!sym.isTrait) {
+          /* Compute the loading spec now, before `flatten` destroys the name
+           * and (in 2.10) the original owner chain. We store it in a global
+           * map.
            */
-          val ownerFullJSName = jsInterop.fullJSNameOf(sym.owner)
-          val jsName = jsInterop.jsNameOf(sym)
-          val fullJSName = ownerFullJSName + "." + jsName
-          sym.addAnnotation(JSFullNameAnnotation,
-              typer.typed(Literal(Constant(fullJSName))))
-        } else if (!sym.isTrait && !sym.hasAnnotation(JSNameAnnotation) &&
-            !isJSGlobalScope(implDef)) {
-          if ((enclosingOwner is OwnerKind.ScalaMod) &&
-              !sym.owner.isPackageObjectClass) {
-            if (sym.isModuleClass) {
-              reporter.error(implDef.pos, "Native JS objects inside " +
-                  "non-native objects must have an @JSName annotation")
+          val loadSpec = {
+            if (enclosingOwner is OwnerKind.JSNativeMod) {
+              val ownerLoadSpec = jsInterop.jsNativeLoadSpecOf(sym.owner)
+              val jsName = jsInterop.jsNameOf(sym)
```

Comment:
Ah, good to know :) Maybe we should prohibit dots in these names then? (via warning and prohibition in 1.0)